agent/consul/leader_connect.go

```diff
@@ -439,64 +419,103 @@ func (s *Server) initializeSecondaryCA(provider ca.Provider, primaryRoots struct
-		// with the time it was rotated out.
-		var newRoots structs.CARoots
-		for _, r := range oldRoots {
-			newRoot := *r
-			if newRoot.Active {
-				newRoot.Active = false
-				newRoot.RotatedOutAt = time.Now()
-			}
-			if newRoot.ExternalTrustDomain == "" {
-				newRoot.ExternalTrustDomain = config.ClusterID
-			}
-			newRoots = append(newRoots, &newRoot)
-		}
-		newRoots = append(newRoots, newActiveRoot)
+	_, config, err := state.CAConfig(nil)
+	if err != nil {
+		return err
+	}
+	if config == nil {
+		return fmt.Errorf("local CA not initialized yet")
+	}
+	newConf := *config
+	newConf.ClusterID = newActiveRoot.ExternalTrustDomain
 
-		args := &structs.CARequest{
-			Op:     structs.CAOpSetRootsAndConfig,
-			Index:  idx,
-			Roots:  newRoots,
-			Config: &newConf,
-		}
-		resp, err := s.raftApply(structs.ConnectCARequestType, &args)
-		if err != nil {
-			return err
-		}
-		if respErr, ok := resp.(error); ok {
-			return respErr
+	// Persist any state the provider needs us to
+	newConf.State, err = provider.State()
+	if err != nil {
+		return fmt.Errorf("error getting provider state: %v", err)
+	}
+
+	// Copy the root list and append the new active root, updating the old root
+	// with the time it was rotated out.
+	var newRoots structs.CARoots
+	for _, r := range oldRoots {
+		newRoot := *r
+		if newRoot.Active {
+			newRoot.Active = false
+			newRoot.RotatedOutAt = time.Now()
 		}
-		if respOk, ok := resp.(bool); ok && !respOk {
-			return fmt.Errorf("could not atomically update roots and config")
+		if newRoot.ExternalTrustDomain == "" {
+			newRoot.ExternalTrustDomain = config.ClusterID
 		}
+		newRoots = append(newRoots, &newRoot)
+	}
+	newRoots = append(newRoots, newActiveRoot)
+
+	args := &structs.CARequest{
+		Op:     structs.CAOpSetRootsAndConfig,
+		Index:  idx,
+		Roots:  newRoots,
+		Config: &newConf,
+	}
+	resp, err := s.raftApply(structs.ConnectCARequestType, &args)
+	if err != nil {
+		return err
+	}
+	if respErr, ok := resp.(error); ok {
+		return respErr
+	}
+	if respOk, ok := resp.(bool); ok && !respOk {
+		return fmt.Errorf("could not atomically update roots and config")
+	}
 
-		s.logger.Printf("[INFO] connect: updated root certificates from primary datacenter")
+	s.logger.Printf("[INFO] connect: updated root certificates from primary datacenter")
+	return nil
+}
+
+func (s *Server) setupNewIntermediate(provider ca.Provider, newActiveRoot *structs.CARoot) error {
+	csr, err := provider.GenerateIntermediateCSR()
+	if err != nil {
+		return err
 	}
 
-	s.setCAProvider(provider, newActiveRoot)
+	var intermediatePEM string
+	if err := s.forwardDC("ConnectCA.SignIntermediate", s.config.PrimaryDatacenter, s.generateCASignRequest(csr), &intermediatePEM); err != nil {
+		// this is a failure in the primary and shouldn't be capable of erroring out our establishing leadership
+		s.logger.Printf("[WARN] connect: Primary datacenter refused to sign our intermediate CA certificate: %v", err)
+		return nil
```

Comment:
If we would return this error, we would have to compare the messages in order to distinguish it from the other errors in this function returns.

Reply:
We can certainly leave this for now but my assumption was that we _never_ let it error on startup currently because it's so bad for the cluster availability so wouldn't need to chose which errors to squash and which not to. If there are still cases we let it error when it starts up now then fair enough.